lib/Inst/Inst.cpp

```diff
@@ -87,7 +87,10 @@ const std::vector<Inst *> &Inst::orderedOps() const {
 }
 
 std::string ReplacementContext::printInst(Inst *I, llvm::raw_ostream &Out,
-                                          bool printNames) {
+                                          bool printNames, Inst *OrigI) {
+  if (OrigI == nullptr)
+    return printInst(I, Out, printNames, I);
```

Comment:
this seems a little weird...  OrigI has a default value of null, but if this function is called with OrigI == null then it calls itself again using a different value for OrigI? can we think of a better way to do this same thing?

also I'm coming to the opinion that we should mostly avoid default values for parameters. how messy would your patch become if you did this differently?

Reply:
The OrigI parameter is designed for holding a "global state" for the recursive function.

One stupid way of getting rid of the default parameter values is to specify the exact OrigI in each function call. This is a bad option since this brings in a redundant argument to every call of the printInst.

Another way is to create a helper function for printInst which does not has OrigI as parameter. Each call to the helper function will be redirected to the printInst, and pass I as OrigI to the printInst.

Finally, the recursive function can always be rewritten into a non-recursive one by adding a stack. 

John: Which way do you prefer?